Add GitHub Copilot Chat Support

Hawkbawk opened the pull request:
# Summary

This commit implements Github Copilot Chat support within the existing Assistant panel/framework. It required a little bit of trickery and internal API modification, as Copilot doesn't use the same authentication-style as all of the existing providers, opting to use OAuth and a short lived API key instead of a straight API key. All existing Assistant features should work.

Release Notes:
- Added Github Copilot Chat support ([#4673](https://github.com/zed-industries/zed/issues/4673)).

## Screenshots
<img width="1552" alt="A screenshot showing a conversation between a user and Github Copilot Chat within the Zed editor." src="https://github.com/user-attachments/assets/73eaf6a2-792b-4c40-a7fe-f763bd6417d7">

Hawkbawk committed:
Add Github Copilot Chat support

This is one of the top requested features from the community, and for
good reason. Some folks can't use OpenAI, Anthropic, or Ollama for
whatever reason, but they or their work might pay for a Github Copilot
and Copilot Chat subscription.

Note that some changes had to be made to the internal APIs of the
LanguageModelCompletionProvider, due to the way that Copilot Chat
authorizes with it's backend.

Release Notes:
- Users can now use Github Copilot Chat from the Assistant panel.
  Copilot Chat is fully compatible with all existing Assistant panel
  features, such as slash commands and inline completion. Users must
  have an active Copilot Chat subscription associated with whatever
  Github account they choose to login with.

Hawkbawk committed:
Update Copilot Chat Icon

The previous assistant icon just didn't feel right, especially given
that we were redirecting the user to Github itself.

Hawkbawk commented:
@cla-bot check

Hawkbawk committed:
Remove unused code, fix model naming/tokens

I had added a method to the http module that I didn't end up needing, so
it's now been deleted.

Additionally, GPT-3.5 Turbo and GPT-4 have different context windows,
so we should reflect that accurately to the user.

maxdeviant commented on crates/copilot/src/copilot_chat.rs:
```diff
@@ -0,0 +1,260 @@
+}
+
+pub async fn fetch_copilot_oauth_token() -> Result<String, anyhow::Error> {
+    let path = if cfg!(target_os = "linux") {
```
We should be using one of our well-known directory paths here instead of computing it on the fly: https://github.com/zed-industries/zed/blob/a5b8094082ec5c033e976572fb5cf48309e5d6b1/crates/paths/src/paths.rs

Hawkbawk replied:
I moved this logic into the paths package and simplified it a little bit at the same time. Let me know how it looks.

maxdeviant commented on crates/copilot/src/copilot_chat.rs:
```diff
@@ -0,0 +1,260 @@
+
+                let host = hosts.get("github.com").ok_or_else(|| anyhow::anyhow!("No host found for github.com"))?.clone();
+
+                let token = host.get("oauth_token").ok_or_else(|| anyhow::anyhow!("No oauth token found for github.com"))?.as_str().unwrap();
```
```suggestion
                let token = host.get("oauth_token").ok_or_else(|| anyhow::anyhow!("No OAuth token found for github.com"))?.as_str().unwrap();
```

Hawkbawk replied:
Fixed!

maxdeviant commented on crates/completion/src/copilot_chat.rs:
```diff
@@ -0,0 +1,336 @@
+impl Render for AuthenticationPrompt {
+    fn render(&mut self, _cx: &mut ViewContext<Self>) -> impl IntoElement {
+        const LABEL: &str =
+            "To use the assistant panel or inline assistant, you must login to Github Copilot. Your Github account must have an active Copilot Chat subscription.";
```
"GitHub" should always be spelled with an uppercase "H" in user-facing text:

```suggestion
            "To use the assistant panel or inline assistant, you must login to GitHub Copilot. Your GitHub account must have an active Copilot Chat subscription.";
```

Hawkbawk replied:
Ah, I did not know that. Fixed it!

Hawkbawk committed:
Fix linting and spelling errors

These were pointed out by clippy or during review.

maxdeviant commented on crates/paths/src/paths.rs:
```diff
@@ -212,6 +212,20 @@ pub fn copilot_dir() -> &'static PathBuf {
+    static COPILOT_CHAT_CONFIG_DIR: OnceLock<PathBuf> = OnceLock::new();
+
+    COPILOT_CHAT_CONFIG_DIR.get_or_init(|| {
+        if cfg!(target_os = "windows") {
```
We should use `config_dir()` here instead of inventing new directories to use.

```rs
COPILOT_CHAT_CONFIG_DIR.get_or_init(|| {
    config_dir().join("github-copilot").join("hosts.json")
})
```

Hawkbawk replied:
The problem is that `config_dir()` points to Zed's config directory, while Copilot stores it's config in it's own directory, namely `~/.config/github-copilot`.

ramenhost replied:
I'm testing this on Windows, and the path appears to be `~\AppData\Local\github-copilot\hosts.json`. GitHub login in Assistant panel doesn't work because of this.

Hawkbawk replied:
Hmmm, that's unfortunate. You'll likely want to raise an issue or create a pull request yourself so this gets fixed. This PR had already been merged. I can likely get to it in the next few days, though I don't have a Windows machine with me right now to test it properly.

ramenhost replied:
The issue is resolved after correcting the path #15587. Great work on this!

bennetbo commented:
Hey, just a heads up, we want to merge #14929 first which changes the way completion providers work. So this will have to wait until then, but I'm happy to review this in a couple of days.

Hawkbawk commented:
@bennetbo That works for me. I looked through the recent pull requests a few days back and saw that one. It looks like some of the changes that you're making there might make this PR easier too, especially with regards to keeping the auth state and short-lived API tokens in sync with the provider model. I'll keep an eye out for when that gets merged and then rebase my work on top of it. Thanks for looking!

Hawkbawk committed:
Merge branch 'main' of github.com:Hawkbawk/zed into add_copilot_chat_support

Hawkbawk committed:
Port Copilot Chat to new language_model crate

Easier to do this in a separate commit so I have a clean reset point
after fixing merge conflicts, if needed.

Hawkbawk committed:
Ensure Chat state is kept in sync

Because Copilot was initialized after the language_model crate, we were
getting an unwrap panic. I shifted it so Copilot is loaded earlier. It
looks like we already do this with supermaven, so I figure it's likely
alright.

Additionally, polish the UI so the user gets better feedback about the
state of everything.

Hawkbawk committed:
Merge branch 'main' of github.com:Hawkbawk/zed into add_copilot_chat_support

Hawkbawk committed:
Fix clippy errors

Also provide better messages for empty prompts and prompts that don't
end with a user message.

Hawkbawk committed:
Fix unwrap panics

I still don't think this is quite right. If they enable Copilot, I don't
think this will pick that up. It'll do as a reset point though.

Hawkbawk committed:
Switch to fsevent system

Rather than listening to the Copilot language server, we'll just listen
to the file that stores the oauth token and extract the token when
necessary. This removes a lot of the complexity and makes things a bit
easier to parse.

Hawkbawk commented:
@bennetbo This should be in a good spot for review now. I might need some help writing tests, but I've also noticed none of the other language model providers have any, so I'm not sure what you want to do here about that. Let me know if you have any questions about the implementation.
This uses a similar approach as https://github.com/CopilotC-Nvim/CopilotChat.nvim, where we let the Copilot inline completion server generate an OAuth Token, fetch that token from the disk, get a short-lived API token, then use that to get completions from GitHub's API.

bennetbo committed:
Merge branch 'main' into add_copilot_chat_support

bennetbo committed:
cleanup

bennetbo committed:
less unwraps

bennetbo committed:
more cleanup and fix tests

bennetbo committed:
move api_token logic to CopilotChat

bennetbo committed:
remove unused dependency

bennetbo committed:
simplify code

bennetbo commented:
@Hawkbawk This is looking pretty good, thank you for putting in all the work. I did some minor refactoring and moved the api_token logic to the CopilotChat implementation, so that we do not need to expose this to the language model provider, ensuring that we always get a valid api token even when calling the CopilotChat methods from elsewhere. 

I am going to test this a bit more today to see if I find any hiccups.

bennetbo commented:
Worked fine for me yesterday, I'll test it a bit more on nightly today and then it will be part of the preview release tomorrow. Thanks for working on this!

baldwindavid commented:
Thanks for all your work on this @Hawkbawk ! I hadn't used the Assistant up to this point, but looking forward to using it now that this has landed.

und3fined commented:
For fun, VSCode Copilot now use GPT-4o 

<img width="1009" alt="Screenshot 2024-08-02 at 12 27 28 AM" src="https://github.com/user-attachments/assets/0dc4296c-176c-4d1f-8415-0838ef2daf98">
